buffer/src/main/java/io/netty/buffer/PooledSlicedByteBuf.java

```diff
@@ -0,0 +1,190 @@
+
+import io.netty.util.IllegalReferenceCountException;
+import io.netty.util.Recycler;
+import io.netty.util.internal.PlatformDependent;
+
+final class PooledSlicedByteBuf extends SlicedAbstractByteBuf {
+
+    private static final AtomicIntegerFieldUpdater<PooledSlicedByteBuf> refCntUpdater;
+
+    static {
+        AtomicIntegerFieldUpdater<PooledSlicedByteBuf> updater = PlatformDependent
+                .newAtomicIntegerFieldUpdater(PooledSlicedByteBuf.class, "refCnt");
+        if (updater == null) {
+            updater = AtomicIntegerFieldUpdater.newUpdater(PooledSlicedByteBuf.class, "refCnt");
+        }
+        refCntUpdater = updater;
+    }
+
+    private volatile int refCnt = 1;
+
+    private final Recycler.Handle recyclerHandle;
+
+    private static final Recycler<PooledSlicedByteBuf> RECYCLER = new Recycler<PooledSlicedByteBuf>() {
+        @Override
+        protected PooledSlicedByteBuf newObject(Handle handle) {
+            return new PooledSlicedByteBuf(handle);
+        }
+    };
+
+    static PooledSlicedByteBuf newInstance(ByteBuf buffer, int index, int length) {
+        PooledSlicedByteBuf buf = RECYCLER.get();
+        buf.init(buffer, index, length);
+        buf.setRefCnt(1);
+        buffer.retain();
+        return buf;
+    }
+
+    private PooledSlicedByteBuf(Recycler.Handle recyclerHandle) {
+        this.recyclerHandle = recyclerHandle;
+    }
+
+    @Override
+    public ByteBuf duplicate(boolean retain) {
+        return retain ? PooledDuplicatedByteBuf.newInstance(this) : duplicate();
+    }
+
+    @Override
+    public ByteBuf duplicate() {
+        return new DuplicatedAbstractByteBuf(this);
+    }
+
+    @Override
+    public ByteBuf slice(int index, int length, boolean retain) {
+        return retain ? newInstance(this, index, length) : slice(index, length);
+    }
+
+    @Override
+    public ByteBuf slice(boolean retain) {
+        return slice(readerIndex(), readableBytes(), retain);
+    }
+
+    @Override
+    public ByteBuf slice(int index, int length) {
+        checkIndex0(index, length);
+        return new SlicedAbstractByteBuf(this, index, length);
+    }
+
+    @Override
+    public ByteBuf readSlice(int length, boolean retain) {
+        ByteBuf slice = slice(readerIndex(), length, retain);
+        skipBytes(length);
+        return slice;
+    }
+
+    /**
+     * Called once {@link #refCnt()} is equals 0.
+     */
+    private void deallocate() {
+        unwrap().release();
+        RECYCLER.recycle(this, recyclerHandle);
+    }
+
+    @Override
+    public int refCnt() {
+        return refCnt;
+    }
+
+    /**
+     * An unsafe operation intended for use by a subclass that sets the reference count of the buffer directly
+     */
+    protected void setRefCnt(int refCnt) {
+        this.refCnt = refCnt;
+    }
+
+    @Override
+    public ByteBuf retain() {
+        for (;;) {
+            int refCnt = this.refCnt;
+            if (refCnt == 0) {
```

Comment:
not really, there's some complicated hierarchy in this pr to get the slice to be "abstract" bytebuf